doc/adrs/086_provide_the_variables_available_for_each_operation.adoc

```diff
@@ -0,0 +1,171 @@
+
+
+  private CommonVariables() {
+    // Prevent instantiation
+  }
+}
+----
+
+[source,java]
+----
+import static CommonVariables.ENVIRONMENT;
+import static CommonVariables.EDITING_CONTEXT;
+import static CommonVariables.SELF;
+
+public class DiagramVariableProvider implements IVariableProvider {
+
+  private static final Variable SEMANTIC_EDGE_SOURCE = new Variable("semanticEdgeSource", Object.class, "...");
+  private static final Variable SEMANTIC_EDGE_TARGET = new Variable("semanticEdgeTarget", Object.class, "...");
+  private static final Variable GRAPHICAL_EDGE_SOURCE = new Variable("graphicalEdgeSource", Element.class, "...");
+  private static final Variable GRAPHICAL_EDGE_TARGET = new Variable("graphicalEdgeSource", Element.class, "...");
+
+  @Override
+  public List<Variable> getVariables(String operation) {
+    return switch (operation) {
+      case "Diagram#canCreate", "Diagram#label", "Diagram#targetObjectId" -> List.of(ENVIRONMENT, EDITING_CONTEXT, SELF);
+      case "Node#semanticElements" -> List.of(ENVIRONMENT, EDITING_CONTEXT, SELF);
+      case "Edge#shouldRender" -> List.of(ENVIRONMENT, EDITING_CONTEXT, SELF, SEMANTIC_EDGE_SOURCE, SEMANTIC_EDGE_TARGET, GRAPHICAL_EDGE_SOURCE, GRAPHICAL_EDGE_TARGET)
+      default -> List.of();
+    }
+  }
+}
+----
+
+=== Attach operations
+
+After that, we will attach the operation definition to the code responsible for their execution.
+For that, we can use a simple annotation to indicate that some Java handler is used to implement a specific operation:
+
+[source,java]
+----
+@Operation("Node#semanticElements")
+public class SemanticCandidatesProvider implements Function<VariableManager, List<?>> {
+  public List<?> apply(VariableManager variableManager) {
+    return List.<Object>of();
+  }
+}
+----
+
+Using this annotation, we could start validating the variables attached to the operation to ensure that our expectation are matching what is really happening:
+
+[source,java]
+----
+public interface IOperationValidator {
+  void validate(Object object, VariableManager variableManager);
+}
+----
+
+[source,java]
+----
+@Operation("Node#semanticElements")
+public class SemanticCandidatesProvider implements Function<VariableManager, List<?>> {
+  private final IOperationValidator operationValidator;
+
+  public List<?> apply(VariableManager variableManager) {
+    this.operationValidator.validate(this, variableManager);
+
+    return List.<Object>of();
+  }
+}
+----
+
+The operation validator would have for only job to check the `@Operation` annotation on the class and ensure that the variables found at runtime are matching what was expected.
+It could thus log as a warning missing variables and as an info additional variables.
+
+In a similar fashion, we can attach the operation to an expression using an EAnnotation on object of the View DSL to indicate that the feature NodeDescription#semanticCandidatesExpression is linked to the operation `Node#semanticElements`.
```

Comment:
on object of -> "on an object of", OR, "on objects of"

Reply:
Done